include/sisl/fds/concurrent_insert_vector.hpp

```diff
@@ -0,0 +1,119 @@
+// reads concurrently with writer, they can use sisl::ThreadVector. This data structure is provided as a replacement for
+// simplistic cases where insertion and iteration never happen concurrently. As a result it provides better performance
+// than even sisl::ThreadVector and better debuggability.
+//
+// Benchmark shows atleast 10x better performance on more than 4 threads concurrently inserting with mutex.
+//
+template < typename T >
+class ConcurrentInsertVector {
+private:
+    ExitSafeThreadBuffer< std::vector< T >, size_t > tvector_;
+    std::vector< std::vector< T > const* > per_thread_vec_ptrs_;
+
+public:
+    struct iterator {
+        size_t next_thread{0};
+        size_t next_id_in_thread{0};
+        ConcurrentInsertVector const* vec{nullptr};
+
+        iterator(ConcurrentInsertVector const& v) : vec{&v} {}
+        iterator(ConcurrentInsertVector const& v, bool end_iterator) : vec{&v} {
+            if (end_iterator) { next_thread = vec->per_thread_vec_ptrs_.size(); }
+        }
+
+        void operator++() {
+            ++next_id_in_thread;
+            if (next_id_in_thread >= vec->per_thread_vec_ptrs_[next_thread]->size()) {
+                ++next_thread;
+                next_id_in_thread = 0;
+            }
+        }
+
+        bool operator==(iterator const& other) const = default;
+        bool operator!=(iterator const& other) const = default;
+
+        T const& operator*() const { return vec->per_thread_vec_ptrs_[next_thread]->at(next_id_in_thread); }
+        T const* operator->() const { return &(vec->per_thread_vec_ptrs_[next_thread]->at(next_id_in_thread)); }
+    };
+
+    ConcurrentInsertVector() = default;
+    ConcurrentInsertVector(size_t size) : tvector_{size} {}
+    ConcurrentInsertVector(const ConcurrentInsertVector&) = delete;
+    ConcurrentInsertVector(ConcurrentInsertVector&&) noexcept = delete;
+    ConcurrentInsertVector& operator=(const ConcurrentInsertVector&) = delete;
+    ConcurrentInsertVector& operator=(ConcurrentInsertVector&&) noexcept = delete;
+    ~ConcurrentInsertVector() = default;
+
+    template < typename InputType,
+               typename = typename std::enable_if<
+                   std::is_convertible< typename std::decay< InputType >::type, T >::value >::type >
+    void push_back(InputType&& ele) {
+        tvector_->push_back(std::forward< InputType >(ele));
+    }
+
+    template < class... Args >
+    void emplace_back(Args&&... args) {
+        tvector_->emplace_back(std::forward< Args >(args)...);
+    }
+
+    iterator begin() {
+        tvector_.access_all_threads([this](std::vector< T > const* tvec, bool, bool) {
+            if (tvec) { per_thread_vec_ptrs_.push_back(tvec); }
+            return false;
+        });
+        return iterator{*this};
+    }
+
+    iterator end() { return iterator{*this, true /* end_iterator */}; }
+
+    void foreach_entry(auto&& cb) {
+        tvector_.access_all_threads([this, &cb](std::vector< T > const* tvec, bool, bool) {
```

Comment:
This is not std::function, but a lambda and wanted the caller force to provide lambda and that is why it is rvalue. In that case I don't see any problem in capturing them by reference, because callback is never intended to store anywhere.